src/ReportPortal.Shared/Reporter/LogsReporter.cs

```diff
@@ -49,105 +52,115 @@ public class LogsReporter : ILogsReporter
-                    {
-                        // only if parent reporter is successful
-                        if (!_reporter.StartTask.IsFaulted && !_reporter.StartTask.IsCanceled)
-                        {
-                            var requests = GetBufferedLogRequests(batchCapacity: BatchCapacity);
-
-                            if (requests.Count != 0)
-                            {
-                                foreach (var logItemRequest in requests)
-                                {
-                                    _logRequestAmender.Amend(logItemRequest);
-                                }
-
-                                NotifySending(requests);
-
-                                await _requestExecuter
-                                    .ExecuteAsync(async () => _asyncReporting
-                                        ? await _service.AsyncLogItem.CreateAsync(requests.ToArray())
-                                        : await _service.LogItem.CreateAsync(requests.ToArray()),
-                                        null,
-                                        _reporter.StatisticsCounter.LogItemStatisticsCounter,
-                                        $"Sending {requests.Count} log items...")
-                                    .ConfigureAwait(false);
+            _queue.Add(logRequest);
+        }
 
-                                NotifySent(requests.AsReadOnly());
-                            }
-                        }
-                    }
-                    catch (Exception exp)
-                    {
-                        TraceLogger.Error($"Unexpected error occurred while processing buffered log requests. {exp}");
-                    }
-                }, TaskContinuationOptions.PreferFairness).Unwrap();
-            }
+        public void Finish()
+        {
+            _queue.CompleteAdding();
         }
 
         public void Sync()
         {
             try
             {
+                Finish();
+
                 ProcessingTask?.GetAwaiter().GetResult();
             }
             catch
             {
                 // we don't aware of failed requests for sending log messages (for now)
             }
-
         }
 
-        private List<CreateLogItemRequest> GetBufferedLogRequests(int batchCapacity)
+        private async Task ConsumeLogRequests()
         {
-            var requests = new List<CreateLogItemRequest>();
-
-            var batchContainsItemWithAttachment = false;
-
-            lock (_syncObj)
+            try
             {
-                for (int i = 0; i < batchCapacity; i++)
+                foreach (var logRequest in _queue.GetConsumingEnumerable())
                 {
-                    if (_buffer.Count > 0)
+                    if (logRequest.Attach != null)
                     {
-                        var logItemRequest = _buffer.Peek();
+                        await SendLogRequests(new List<CreateLogItemRequest> { logRequest });
```

Comment:
Our client has overloaded version of CreateAsync(...) method which can handle single CreateLogItemRequest. Maybe make sense to re-use it here for avoiding creation new list just for wrapping single request. What do you think?

Reply:
I even tried to use Array instead of List, looks good. But one unit test failed: we have a feature, when user may add/delete items. And it doesn't work with fixed-size collections (Array).

Reply:
Got it!